### PR TITLE
Introduce static_to_shared_library function

### PR DESCRIPTION
**Pull request by michaelkuhn**
This is work in progress based on the discussion in #6008.

@junghans: Sorry for the timing. I was just working on this when you merged #6008. :-)

The `static_to_shared_library` function takes an existing static library and produces a shared library based on it.

For now, I just put it into the lua package. I would like to get some feedback whether the function's interface etc. are what you had in mind. Additionally, I would need some hints on how to integrate this better into Spack. (I have already experimented with putting it into `build_environment.py` but I currently need access to `spack_cc` and `dso_suffix`.)

### CODE REVIEW

**Comment by junghans** [comment]
@tgamblin I would really like to see `static_to_shared_library` as a part of the core, where would we put this?

**Comment by michaelkuhn** [comment]
Small update: I moved it into the package class, which has the benefit of giving us access to the spec.

**Comment by michaelkuhn** [comment]
I have refactored the code a bit, put it into `build_environment.py` and added a rough first test for it (see below).

> How well does this work across different compilers? Are the arguments portable?

From what I could find, at least GCC, Intel and Clang should support `-Wl`. GCC and Clang support `-shared`, `-dynamiclib` and `-install_name`. The rest depends on the linker.

> Does this require that the static lib already be PIC? Or does it work in general?

The static library has to be compiled with PIC.

> This seems like something that should only exist in the build environment (though if there are other uses maybe we should rethink that). Maybe the right place for it is in build_environment.py, and it could be a function added to the module scope in, e.g., set_module_variables_for_package(). You have the Package and spack_cc there, so you could construct a converter function specialized for the build.

Done. I moved the compiler selection out of the function to be able to test it more easily.

> Needs a test.

I added a rough first test for it that can be extended later. Please check whether this is what you had in mind.

**Comment by mathstuf** [comment]
Open questions:

  - What to do for Windows? Symbols need to be exported there, so a `.def` file listing the symbols to export from the shared library is required. Headers may also need patched to set the proper `dllimport` settings (static libraries tend to have no `declspec` at all). Blindly listing the symbols for a `.def` file may end up exporting supposed-to-be-internal symbols as well.
  - How to handle multiple static libraries in a package? What if they have circular dependencies?
  - Static builds may have compiled out some features necessary in a shared build (though it seems this is more for coercing never-shared projects to be shared rather than replacing an intentional `link=shared` option).
  - Are soname numbers just skipped? What about library IDs on macOS? I see something is set; is that a full path or just the filename?
  - Other compiler support?

**Comment by michaelkuhn** [comment]
> What to do for Windows? Symbols need to be exported there, so a .def file listing the symbols to export from the shared library is required. Headers may also need patched to set the proper dllimport settings (static libraries tend to have no declspec at all). Blindly listing the symbols for a .def file may end up exporting supposed-to-be-internal symbols as well.

No idea, to be honest. Someone else will have to take care of this, as I do not have access to a Windows build environment.

> How to handle multiple static libraries in a package? What if they have circular dependencies?
Static builds may have compiled out some features necessary in a shared build (though it seems this is more for coercing never-shared projects to be shared rather than replacing an intentional link=shared option).

This is just meant as an easy way to convert existing static libraries to shared ones, in case the package in question does not build them (see lua). Instead of patching the package, we can just use this function.

> Are soname numbers just skipped? What about library IDs on macOS? I see something is set; is that a full path or just the filename?

Sorry, I skipped them for the initial version. I just pushed an updated version with support for specifying a version.

> Other compiler support?

I think it should be pretty easy to extend the function to support additional compilers.

File: lib/spack/spack/test/build_environment.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        os.environ['SPACK_FC_RPATH_ARG']  = "-Wl,-rpath,"
+
+        if 'SPACK_DEPENDENCIES' in os.environ:
+            del os.environ['SPACK_DEPENDENCIES']
+
+    def test_static_to_shared_library(self):
+        os.environ['SPACK_TEST_COMMAND'] = 'dump-args'
+
+        expected = {
+            'linux': '/bin/mycc -Wl,-rpath,/spack-test-prefix/lib -Wl,-rpath,/spack-test-prefix/lib64 -shared -Wl,-soname,{2} -Wl,--whole-archive {0} -Wl,--no-whole-archive -o {1}',  # noqa: E501
```
Comment (tgamblin) [inline lib/spack/spack/test/build_environment.py]: split across multiple lines (no NOQA)

File: lib/spack/spack/test/build_environment.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+import spack
+from llnl.util.filesystem import join_path
+from spack.build_environment import dso_suffix, _static_to_shared_library
+from spack.util.executable import Executable
+
+
+class BuildEnvironmentTest(unittest.TestCase):
+
+    def setUp(self):
```
Comment (tgamblin) [inline lib/spack/spack/test/build_environment.py]: can you use a `pytest` fixture for this?  We still have some old `unittest` style tests but we're trying to switch to `pytest`.  You can probably just make a `build_environment` fixture that does all the same stuff as this, and yields a named tuple with `cc`, `ld`, etc.  It should probably also unset the environment variables after yielding so as not to interfere with other tests.

File: lib/spack/spack/test/build_environment.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+        os.environ['SPACK_CC_RPATH_ARG']  = "-Wl,-rpath,"
+        os.environ['SPACK_CXX_RPATH_ARG'] = "-Wl,-rpath,"
+        os.environ['SPACK_F77_RPATH_ARG'] = "-Wl,-rpath,"
+        os.environ['SPACK_FC_RPATH_ARG']  = "-Wl,-rpath,"
+
+        if 'SPACK_DEPENDENCIES' in os.environ:
+            del os.environ['SPACK_DEPENDENCIES']
+
+    def test_static_to_shared_library(self):
```
Comment (tgamblin) [inline lib/spack/spack/test/build_environment.py]: This test is fine but can be moved to module scope if it's a pytest test.  It should take the `build_environment` fixture as a parameter.

File: lib/spack/spack/build_environment.py [hunk 4]
```diff
@@ -371,6 +371,72 @@ def set_module_variables_for_package(pkg, module):
...
+    compiler_args = []
+
+    if 'linux' in arch:
+        soname = os.path.basename(shared_lib)
+
+        if version:
+            soname += '.{0}'.format(version)
+
+        compiler_args = [
+            '-shared',
```
Comment (tgamblin) [inline lib/spack/spack/build_environment.py]: can you add a TODO in here somewhere that says these arguments should really be added to the various compiler classes and generalized for different compilers?

File: lib/spack/spack/build_environment.py [hunk 5]
```diff
@@ -371,6 +371,72 @@ def set_module_variables_for_package(pkg, module):
...
+    m.static_to_shared_library = static_to_shared_library
+
+
+def _static_to_shared_library(arch, compiler, static_lib, shared_lib=None,
+                              **kwargs):
+    """
+    Converts a static library to a shared library.
+
+    Parameters:
+        static_lib (str): Path to the static library.
```
Comment (tgamblin) [inline lib/spack/spack/build_environment.py]: There should be a note here or in the description that this library *must* be built with PIC.

**Comment by michaelkuhn** [comment]
I have pushed a new version. It includes some updates regarding the version handling (if we set a version `X.Y.Z`, the library is expected to be found at `libfoo.so.X.Y.Z`, so we create a few symlinks) and should address all of @tgamblin's comments.

**Comment by tgamblin** [comment]
Thanks!  LGTM!

**Comment by michaelkuhn** [comment]
I wonder why the CI tests are failing, they work fine locally. Is there something special about the CI environment?

File: lib/spack/spack/test/build_environment.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
+##############################################################################
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
```
Comment (junghans) [inline lib/spack/spack/test/build_environment.py]: Update url to `spack/spack`!

**Comment by michaelkuhn** [comment]
OK, I think I have found the problem (`SPACK_DEBUG_LOG_ID` was missing). Should be good to go now. :-)

**Pull request merged**